rasqc/utils.py

```diff
@@ -65,3 +74,146 @@ def flatten(group_name):
+    """Convert a python object to an HTML preformatted string."""
+    return BeautifulSoup(
+        f"<style>pre {{color: {color_code}}}</style>\n<pre>{remove_json_seps(json.dumps(obj, indent=4))}</pre>",
+        features="html.parser",
+    ).prettify()
+
+
+def pyobj_to_html_table(obj, color_code: str = "rgb(170, 170, 170)") -> str:
+    """Convert a python object to an HTML table string."""
+    html = f'<table border="0" style="line-height:1em; border-spacing:0; color:{color_code};">'
+    if isinstance(obj, dict):
+        for key, value in obj.items():
+            html += "<tr>"
+            html += f"<th valign='top' align='left' style='white-space: nowrap;'>{key} :</th>"
+            html += f"<td valign='top' align='left'>{pyobj_to_html_table(value, color_code)}</td>"
+            html += "</tr>"
+    elif isinstance(obj, (list, set, tuple)):
+        for item in obj:
+            html += "<tr>"
+            html += f"<td valign='top' align='left'>{pyobj_to_html_table(item, color_code)}</td>"
+            html += "</tr>"
+    else:
+        html += f"<td valign='top' align='left'>{obj}</td>"
+    html += "</table>"
+    return BeautifulSoup(html, features="html.parser").prettify()
+
+
+def is_valid_json(json_str: str) -> bool:
+    """Check if string is a valid JSON string."""
+    try:
+        json.loads(json_str)
+        return True
+    except ValueError:
+        return False
+
+
+def group_results(results: List[RasqcResult]) -> dict:
+    """Group RasqcResult objects by result status.
+
+    Group a list of RasqcResult objects into a dict based on
+    whether the result is from a 'note' or 'check', then based
+    on the name of the RasqcResult object.
+
+    Parameters
+    ----------
+        results: A list of RasqcResult objects.
+
+    Returns
+    -------
+        dict: A dict of RasqcResult objects in pattern {'note' or 'check': {result name: [results]}}.
+    """
+    results_dict = {}
+    for result in results:
+        results_dict.setdefault(
+            "note" if result.result.value == "note" else "check", {}
+        ).setdefault(result.name, []).append(result)
+    return results_dict
+
+
+def results_to_html(
+    results: list[RasqcResult],
+    output_path: str,
+    model_path: str,
+    checksuite: str,
+    message_style: Literal["table", "indent"] = "table",
+    tool_version: str = None,
+    theme: ColorTheme = ColorTheme.ARCADE,
+) -> None:
+    """Create an HTML file from a list of RasqcResults."""
+    env = Environment(loader=FileSystemLoader(Path(__file__).parent.resolve()))
+    env.globals.update(
+        pyobj_to_html_table=pyobj_to_html_table,
+        pyobj_to_html_string=pyobj_to_html_string,
+        message_style=message_style,
+        is_valid_json=is_valid_json,
+        loads=json.loads,
+    )
+    template = env.get_template("template.html")
+    details = (
+        subprocess.check_output(["echo", "%USERNAME%", "%DATE%", "%TIME%"], shell=True)
```

Comment:
This fails on non-Windows systems since Linux/MacOS environment variables shell works differently. But there's a platform-independent way to get the username and datetime:
```python
from datetime import datetime
from getpass import getuser

user = getuser()
dt = datetime.now()
```